twoliter/src/project/migrate/mod.rs

```diff
@@ -0,0 +1,553 @@
+                        current_version,
+                        next_version
+                    );
+                }
+            }
+        }
+
+        Ok(current_content)
+    }
+}
+
+/// Detect the schema version of a project file without fully parsing it.
+pub fn detect_schema_version(content: &str) -> Result<u32> {
+    // Parse just enough to get the schema version
+    let value: toml::Value = toml::from_str(content).context("Failed to parse TOML content")?;
+
+    let version = value
+        .get("schema-version")
+        .or_else(|| value.get("schema_version"))
+        .context("No schema-version field found in project file")?
+        .as_integer()
+        .context("schema-version field is not an integer")?;
+
+    Ok(version as u32)
+}
+
+/// Migrate project content to the target version.
+/// This is the main function used by the project loading system to automatically
+/// handle migration when loading Twoliter.toml files.
+pub fn migrate_project_content(content: &str, target_version: u32) -> Result<String> {
+    let current_version =
+        detect_schema_version(content).context("Failed to detect schema version")?;
+
+    if current_version == target_version {
+        debug!(
+            "Project file is already at target version {}",
+            target_version
+        );
+        return Ok(content.to_string());
+    }
+
+    let registry = MigrationRegistry::new();
+
+    if !registry.can_migrate(current_version, target_version) {
+        if current_version > target_version {
+            warn!(
+                "Project file version ({}) is newer than target version ({}). No backward migration available.",
+                current_version, target_version
+            );
+        } else {
+            warn!(
+                "Project file version ({}) is older than target version ({}). No forward migration available.",
+                current_version, target_version
+            );
+        }
+        return Ok(content.to_string());
+    }
+
+    registry.migrate_project(content, current_version, target_version)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    // Mock migrator for testing
+    struct MockMigrator {
+        from: u32,
+        to: u32,
+        transform: fn(&str) -> Result<String>,
+    }
+
+    impl MockMigrator {
+        fn new(from: u32, to: u32, transform: fn(&str) -> Result<String>) -> Self {
+            Self {
+                from,
+                to,
+                transform,
+            }
+        }
+    }
+
+    impl ProjectMigrator for MockMigrator {
+        fn current_version(&self) -> u32 {
+            self.from
+        }
+
+        fn to_version(&self) -> u32 {
+            self.to
+        }
+
+        fn migrate(&self, input: &str) -> Result<String> {
+            (self.transform)(input)
+        }
+    }
+
+    #[test]
+    fn test_project_migrator_can_migrate() {
+        let migrator = MockMigrator::new(1, 2, |s| Ok(s.to_string()));
```

Comment:
Given the bug identified above, we should expand this test.

Consider using the `test_case` macro here, it lets you defined parameterized test cases on a given test function. There are other examples of its usage in the Twoliter repository.